File: tests/language/function_subtype0_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Dart test program for constructors and initializers.

// Check function subtyping.

import 'package:expect/expect.dart';

typedef t__();
typedef void t_void_();
typedef void t_void_2();
typedef int t_int_();
typedef int t_int_2();
typedef Object t_Object_();
typedef double t_double_();
typedef void t_void__int(int i);
typedef int t_int__int(int i);
typedef int t_int__int2(int i);
typedef int t_int__Object(Object o);
typedef Object t_Object__int(int i);
typedef int t_int__double(double d);
typedef int t_int__int_int(int i1, int i2);
typedef void t_inline_void_(void f());
typedef void t_inline_void__int(void f(int i));

void _() => null;
void void_() {}
void void_2() {}
int int_() => 0;
int int_2() => 0;
Object Object_() => null;
double double_() => 0.0;
void void__int(int i) {}
int int__int(int i) => 0;
int int__int2(int i) => 0;
int int__Object(Object o) => 0;
Object Object__int(int i) => null;
int int__double(double d) => 0;
int int__int_int(int i1, int i2) => 0;
void inline_void_(void f()) {}
void inline_void__int(void f(int i)) {}

main() {
  // () -> int <: Function
  Expect.isTrue(int_ is Function);
  // () -> dynamic <: () -> dynamic
  Expect.isTrue(_ is t__);
  // () -> dynamic <: () -> void
  Expect.isTrue(_ is t_void_);
  // () -> void <: () -> dynamic
  Expect.isTrue(void_ is t__);
  // () -> int <: () -> void
  Expect.isTrue(int_ is t_void_);
  // () -> void <: () -> int
  Expect.isFalse(void_ is t_int_);
  // () -> void <: () -> void
  Expect.isTrue(void_ is t_void_2);
  // () -> int <: () -> int
  Expect.isTrue(int_ is t_int_2);
  // () -> int <: () -> Object
  Expect.isTrue(int_ is t_Object_);
  // () -> int <: () -> double
  Expect.isFalse(int_ is t_double_);
  // () -> int <: (int) -> void
  Expect.isFalse(int_ is t_void__int);
  // () -> void <: (int) -> int
  Expect.isFalse(void_ is t_int__int);
  // () -> void <: (int) -> void
  Expect.isFalse(void_ is t_void__int);
  // (int) -> int <: (int) -> int
  Expect.isTrue(int__int is t_int__int2);
  // (Object) -> int <: (int) -> Object
  Expect.isTrue(int__Object is t_Object__int);
  // (int) -> int <: (double) -> int
  Expect.isFalse(int__int is t_int__double);
  // () -> int <: (int) -> int
  Expect.isFalse(int_ is t_int__int);
  // (int) -> int <: (int,int) -> int
  Expect.isFalse(int__int is t_int__int_int);
  // (int,int) -> int <: (int) -> int
  Expect.isFalse(int__int_int is t_int__int);
  // (()->void) -> void <: ((int)->void) -> void
  Expect.isFalse(inline_void_ is t_inline_void__int);
  // ((int)->void) -> void <: (()->void) -> void
  Expect.isFalse(inline_void__int is t_inline_void_);
}

File: pkg/dev_compiler/test/codegen_expected/language/function_subtype0_test.js
dart_library.library('language/function_subtype0_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__function_subtype0_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const function_subtype0_test = Object.create(null);
  let VoidTovoid = () => (VoidTovoid = dart.constFn(dart.functionType(dart.void, [])))();
  let intTovoid = () => (intTovoid = dart.constFn(dart.functionType(dart.void, [core.int])))();
  let VoidTovoid$ = () => (VoidTovoid$ = dart.constFn(dart.definiteFunctionType(dart.void, [])))();
  let VoidToint = () => (VoidToint = dart.constFn(dart.definiteFunctionType(core.int, [])))();
  let VoidToObject = () => (VoidToObject = dart.constFn(dart.definiteFunctionType(core.Object, [])))();
  let VoidTodouble = () => (VoidTodouble = dart.constFn(dart.definiteFunctionType(core.double, [])))();
  let intTovoid$ = () => (intTovoid$ = dart.constFn(dart.definiteFunctionType(dart.void, [core.int])))();
  let intToint = () => (intToint = dart.constFn(dart.definiteFunctionType(core.int, [core.int])))();
  let ObjectToint = () => (ObjectToint = dart.constFn(dart.definiteFunctionType(core.int, [core.Object])))();
  let intToObject = () => (intToObject = dart.constFn(dart.definiteFunctionType(core.Object, [core.int])))();
  let doubleToint = () => (doubleToint = dart.constFn(dart.definiteFunctionType(core.int, [core.double])))();
  let intAndintToint = () => (intAndintToint = dart.constFn(dart.definiteFunctionType(core.int, [core.int, core.int])))();
  let FnTovoid = () => (FnTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [VoidTovoid()])))();
  let FnTovoid$ = () => (FnTovoid$ = dart.constFn(dart.definiteFunctionType(dart.void, [intTovoid()])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  function_subtype0_test.t__ = dart.typedef('t__', () => dart.functionType(dart.dynamic, []));
  function_subtype0_test.t_void_ = dart.typedef('t_void_', () => dart.functionType(dart.void, []));
  function_subtype0_test.t_void_2 = dart.typedef('t_void_2', () => dart.functionType(dart.void, []));
  function_subtype0_test.t_int_ = dart.typedef('t_int_', () => dart.functionType(core.int, []));
  function_subtype0_test.t_int_2 = dart.typedef('t_int_2', () => dart.functionType(core.int, []));
  function_subtype0_test.t_Object_ = dart.typedef('t_Object_', () => dart.functionType(core.Object, []));
  function_subtype0_test.t_double_ = dart.typedef('t_double_', () => dart.functionType(core.double, []));
  function_subtype0_test.t_void__int = dart.typedef('t_void__int', () => dart.functionType(dart.void, [core.int]));
  function_subtype0_test.t_int__int = dart.typedef('t_int__int', () => dart.functionType(core.int, [core.int]));
  function_subtype0_test.t_int__int2 = dart.typedef('t_int__int2', () => dart.functionType(core.int, [core.int]));
  function_subtype0_test.t_int__Object = dart.typedef('t_int__Object', () => dart.functionType(core.int, [core.Object]));
  function_subtype0_test.t_Object__int = dart.typedef('t_Object__int', () => dart.functionType(core.Object, [core.int]));
  function_subtype0_test.t_int__double = dart.typedef('t_int__double', () => dart.functionType(core.int, [core.double]));
  function_subtype0_test.t_int__int_int = dart.typedef('t_int__int_int', () => dart.functionType(core.int, [core.int, core.int]));
  function_subtype0_test.t_inline_void_ = dart.typedef('t_inline_void_', () => dart.functionType(dart.void, [dart.functionType(dart.void, [])]));
  function_subtype0_test.t_inline_void__int = dart.typedef('t_inline_void__int', () => dart.functionType(dart.void, [dart.functionType(dart.void, [core.int])]));
  function_subtype0_test._ = function() {
    return null;
  };
  dart.fn(function_subtype0_test._, VoidTovoid$());
  function_subtype0_test.void_ = function() {
  };
  dart.fn(function_subtype0_test.void_, VoidTovoid$());
  function_subtype0_test.void_2 = function() {
  };
  dart.fn(function_subtype0_test.void_2, VoidTovoid$());
  function_subtype0_test.int_ = function() {
    return 0;
  };
  dart.fn(function_subtype0_test.int_, VoidToint());
  function_subtype0_test.int_2 = function() {
    return 0;
  };
  dart.fn(function_subtype0_test.int_2, VoidToint());
  function_subtype0_test.Object_ = function() {
    return null;
  };
  dart.fn(function_subtype0_test.Object_, VoidToObject());
  function_subtype0_test.double_ = function() {
    return 0.0;
  };
  dart.fn(function_subtype0_test.double_, VoidTodouble());
  function_subtype0_test.void__int = function(i) {
  };
  dart.fn(function_subtype0_test.void__int, intTovoid$());
  function_subtype0_test.int__int = function(i) {
    return 0;
  };
  dart.fn(function_subtype0_test.int__int, intToint());
  function_subtype0_test.int__int2 = function(i) {
    return 0;
  };
  dart.fn(function_subtype0_test.int__int2, intToint());
  function_subtype0_test.int__Object = function(o) {
    return 0;
  };
  dart.fn(function_subtype0_test.int__Object, ObjectToint());
  function_subtype0_test.Object__int = function(i) {
    return null;
  };
  dart.fn(function_subtype0_test.Object__int, intToObject());
  function_subtype0_test.int__double = function(d) {
    return 0;
  };
  dart.fn(function_subtype0_test.int__double, doubleToint());
  function_subtype0_test.int__int_int = function(i1, i2) {
    return 0;
  };
  dart.fn(function_subtype0_test.int__int_int, intAndintToint());
  function_subtype0_test.inline_void_ = function(f) {
  };
  dart.fn(function_subtype0_test.inline_void_, FnTovoid());
  function_subtype0_test.inline_void__int = function(f) {
  };
  dart.fn(function_subtype0_test.inline_void__int, FnTovoid$());
  function_subtype0_test.main = function() {
    expect$.Expect.isTrue(core.Function.is(function_subtype0_test.int_));
    expect$.Expect.isTrue(function_subtype0_test.t__.is(function_subtype0_test._));
    expect$.Expect.isTrue(function_subtype0_test.t_void_.is(function_subtype0_test._));
    expect$.Expect.isTrue(function_subtype0_test.t__.is(function_subtype0_test.void_));
    expect$.Expect.isTrue(function_subtype0_test.t_void_.is(function_subtype0_test.int_));
    expect$.Expect.isFalse(function_subtype0_test.t_int_.is(function_subtype0_test.void_));
    expect$.Expect.isTrue(function_subtype0_test.t_void_2.is(function_subtype0_test.void_));
    expect$.Expect.isTrue(function_subtype0_test.t_int_2.is(function_subtype0_test.int_));
    expect$.Expect.isTrue(function_subtype0_test.t_Object_.is(function_subtype0_test.int_));
    expect$.Expect.isFalse(function_subtype0_test.t_double_.is(function_subtype0_test.int_));
    expect$.Expect.isFalse(function_subtype0_test.t_void__int.is(function_subtype0_test.int_));
    expect$.Expect.isFalse(function_subtype0_test.t_int__int.is(function_subtype0_test.void_));
    expect$.Expect.isFalse(function_subtype0_test.t_void__int.is(function_subtype0_test.void_));
    expect$.Expect.isTrue(function_subtype0_test.t_int__int2.is(function_subtype0_test.int__int));
    expect$.Expect.isTrue(function_subtype0_test.t_Object__int.is(function_subtype0_test.int__Object));
    expect$.Expect.isFalse(function_subtype0_test.t_int__double.is(function_subtype0_test.int__int));
    expect$.Expect.isFalse(function_subtype0_test.t_int__int.is(function_subtype0_test.int_));
    expect$.Expect.isFalse(function_subtype0_test.t_int__int_int.is(function_subtype0_test.int__int));
    expect$.Expect.isFalse(function_subtype0_test.t_int__int.is(function_subtype0_test.int__int_int));
    expect$.Expect.isFalse(function_subtype0_test.t_inline_void__int.is(function_subtype0_test.inline_void_));
    expect$.Expect.isFalse(function_subtype0_test.t_inline_void_.is(function_subtype0_test.inline_void__int));
  };
  dart.fn(function_subtype0_test.main, VoidTodynamic());
  // Exports:
  exports.function_subtype0_test = function_subtype0_test;
});
